src/Nervos/Address.cpp

```diff
@@ -0,0 +1,160 @@
+            }
+            if (decodedPayload.size() != 22) {
+                break;
+            }
+            codeHashIndex = decodedPayload[1];
+            if (codeHashIndex != 0) {
+                break;
+            }
+            codeHash = CommonFunc::getSecp256k1CodeHash();
+            hashType = HashType::Type1;
+            args = Data(decodedPayload.begin() + 2, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        case AddressType::DataCodeHash:
+        case AddressType::TypeCodeHash: {
+            if (decodedVariant != Bech32::ChecksumVariant::Bech32) {
+                break;
+            }
+            if (decodedPayload.size() < 33) {
+                break;
+            }
+            codeHashIndex = -1;
+            codeHash = Data(decodedPayload.begin() + 1, decodedPayload.begin() + 33);
+            hashType = addressType == AddressType::DataCodeHash ? HashType::Data0 : HashType::Type1;
+            args = Data(decodedPayload.begin() + 33, decodedPayload.end());
+            isValid = true;
+            break;
+        }
+        }
+    } while (false);
+
+    if (!isValid) {
+        throw std::invalid_argument("Invalid address string");
+    }
+}
+
+Address::Address(const PublicKey& publicKey)
+    : Address(publicKey, TW::p2pkhPrefix(TWCoinTypeNervos),
+              stringForHRP(TW::hrp(TWCoinTypeNervos))) {}
+
+Address::Address(const PublicKey& publicKey, byte p2pkh, const char* hrp) : hrp(hrp) {
+    if (publicKey.type != TWPublicKeyTypeSECP256k1) {
+        throw std::invalid_argument("Nervos::Address needs a SECP256k1 public key.");
+    }
+    addressType = AddressType::FullVersion;
+    codeHashIndex = -1;
+    codeHash = CommonFunc::getSecp256k1CodeHash();
+    hashType = HashType::Type1;
+    args = TW::Hash::blake2b(publicKey.bytes, 32, CommonFunc::getHashPersonalization());
+    args = Data(args.begin(), args.begin() + 20);
+}
+
+std::string Address::string() const {
+    auto data = Data();
+    data.push_back((byte)addressType);
+    Bech32::ChecksumVariant checksumVariant;
+    switch (addressType) {
+    case AddressType::FullVersion: {
+        data.insert(data.end(), codeHash.begin(), codeHash.end());
+        data.push_back((byte)hashType);
+        data.insert(data.end(), args.begin(), args.end());
+        checksumVariant = Bech32::ChecksumVariant::Bech32M;
+        break;
+    }
+    case AddressType::HashIdx: {
+        data.push_back(codeHashIndex);
+        data.insert(data.end(), args.begin(), args.end());
+        checksumVariant = Bech32::ChecksumVariant::Bech32;
+        break;
+    }
+    case AddressType::DataCodeHash:
+    case AddressType::TypeCodeHash: {
+        data.insert(data.end(), codeHash.begin(), codeHash.end());
+        data.insert(data.end(), args.begin(), args.end());
+        checksumVariant = Bech32::ChecksumVariant::Bech32;
+        break;
+    }
+    default: {
+        throw std::invalid_argument("Invalid address string");
```

Comment:
This throw is not justified, the function doesn't take any argument, consider using a more appropriate exception

Reply:
Done